text/0000-route-prefix.md

```diff
@@ -0,0 +1,109 @@
+
+```
+website:
+  command: 'ROUTE_PREFIX="/privacy" FRAMEWORK_STATIC_ASSET_PATH="https://someuuid.cloudfront.net/uberprivacy" npm start'
+```
+
+Ideally, as this is being provided as an environment variable, we should scope this to the `fusion-cli` and ensure it is not exposed to user land.
+
+Note that the primary [webpack considerations](https://github.com/fusionjs/fusion-cli/search?utf8=%E2%9C%93&q=__webpack_public_path__&type=) here are already handled via the [`compilation-metadata-plugin`](https://github.com/fusionjs/fusion-cli/blob/master/plugins/compilation-metadata-plugin.js) plugin.
+
+### `ctx.url` rewrite
+
+Many middleware are expected to run conditional checks against `ctx.url`.  It is not the concern of the middleware to know whether the web app is deployed at the root of a host, or behind a route prefix.  In order to enforce this, a core middleware can strip the route prefix from `ctx.url` prior to any of the user-land middlewares' execution.  This has the benefit of allowing middlewares to be easily re-used across applications regardless of deployment concerns.
+
+e.g.
+```
+# fusion-cli: src/route-prefix-plugin.js
+const envVarsPlugin = require('./environment-variables-plugin');
+module.exports = function() {
+  const envVars = envVarsPlugin().of();
+  return function middleware(ctx, next) {
+    const prefix = envVars.prefix;
+
+    // enhance ctx.url sans prefix
+    if(ctx.url.indexOf(prefix) === 0 /*found at index 0*/) {
+      ctx.url = ctx.url.slice(prefix.length);
+    }
+  }
+}
+```
+
+### Prefix handling for `fusion-plugin-react-router`
+
+In order to ensure that the router works as expected, we will want to pass in the prefix as part of the [configuration](https://github.com/fusionjs/fusion-plugin-react-router#router).
+
+Today, the router has the option to take in a `basepath` as a prop:
+
+```
+import {Router} from 'fusion-plugin-react-router';
+
+<Router
+  location={...}
+  basename={...}
+  context={...}
+  onRoute={...}
+>{child}</Router>
+```
+[(link)](https://github.com/fusionjs/fusion-plugin-react-router#router)
+
+As user land plugins should not be accessing environment variables directly, this will likely be supplied by a user land `{config}.js` file.  This file must remain synchronized to the configuration used in the static asset plugin and the ctx re-write.  Due to this being sensitive to errors, it may be prudent to have a build time check here to ensure synchronization.
+
+# Drawbacks
+
+This is a non-trival amount of effort to implement.  In its current form, it is also prone to user error and requires duplicated configuration.  As noted below, the alternative in the [dependency injection](https://github.com/fusionjs/rfcs/pull/1/commits/8c440259291c9774b44b9dca07cf35dbf2a5f98e?short_path=9919d4f#diff-9919d4f45f05666e94a1bcb5edd7f51a) world is potentially cleaner.  Implementing the proposal above will require a migration path towards the DI implementation down the road.
```

Comment:
Can you expand on which files would be duplicates? The yaml file with the basepath definition and what else? I probably missed the other place.

Reply:
Sure!  At a high level, the two places would likely be `pinocchio.yaml` and a user land config file, like `routePrefixConfig.js`.

* The yaml file would power the environment variables used by `fusion-cli` which includes populating [webpack's public path](https://github.com/fusionjs/fusion-cli/search?utf8=%E2%9C%93&q=__webpack_public_path__&type=) and [`assetUrl`](https://github.com/fusionjs/fusion-cli/blob/af9da1ff34be8c9546f6de5c75b87850024d1e1a/build/babel-plugins/babel-plugin-asseturl/index.js).
* The user land configuration would be used to configure user land plugins - which should not access environment variables - like [`fusion-plugin-react-router`](https://github.com/fusionjs/fusion-plugin-react-router#router) and [`fusion-plugin-font-loader-react`](https://github.com/fusionjs/fusion-plugin-font-loader-react/search?utf8=%E2%9C%93&q=_static&type=).